src/prune.ts

```diff
@@ -0,0 +1,251 @@
+              project,
+              masterDependencies,
+              dtsPath: join(typingsPath, DTS_BROWSER_FILE),
+              emitter
+            }),
+            // Prune main.d.ts
+            pruneExtraneousReferences({
+              project,
+              masterDependencies,
+              dtsPath: join(typingsPath, DTS_MAIN_FILE),
+              emitter
+            }),
+            // Prune typings/browser folder
+            pruneExtraneousTypings({
+              masterDependencies,
+              typingsPath: join(typingsPath, BROWSER_TYPINGS_DIR),
+              emitter
+            }),
+            // Prune typings/main folder
+            pruneExtraneousTypings({
+              masterDependencies,
+              typingsPath: join(typingsPath, MAIN_TYPINGS_DIR),
+              emitter
+            }),
+          ]).then(() => { }) // convert to Promise<void>
+        })
+    })
+}
+
+interface TypeDependencies {
+  ambientDependencies: string[]
+  externalDependencies: string[]
+}
+
+function readMasterDependencies(projectDirectory: string, includeDev: boolean): Promise<TypeDependencies> {
+  return readConfig(join(projectDirectory, CONFIG_FILE))
+    .then((config) => {
+      const ambientDependencies = config.ambientDependencies != null
+        ? Object.keys(config.ambientDependencies)
+        : []
+
+      const externalDependencies = config.dependencies != null
+        ? Object.keys(config.dependencies)
+        : []
+
+      const dependencies: TypeDependencies = {
+        ambientDependencies,
+        externalDependencies
+      }
+
+      if (includeDev) {
+        const ambientDevDependencies = config.ambientDevDependencies != null
+          ? Object.keys(config.ambientDevDependencies)
+          : []
+
+        const externalDevDependencies = config.devDependencies != null
+          ? Object.keys(config.devDependencies)
+          : []
+
+        dependencies.ambientDependencies = dependencies.ambientDependencies.concat(ambientDevDependencies)
+        dependencies.externalDependencies = dependencies.externalDependencies.concat(externalDevDependencies)
+      }
+
+      return dependencies
+    })
+}
+
+interface PruneReferencesOptions {
+  masterDependencies: TypeDependencies
+  project: string
+  dtsPath: string
+  emitter: Emitter
+}
+
+function pruneExtraneousReferences(options: PruneReferencesOptions): Promise<void> {
+  const { masterDependencies, dtsPath, emitter } = options
+
+  return readFile(dtsPath, 'utf8')
+    .then((config: string) => {
+      const lines = config.split('\n')
+      // Split into extraneous and non-extraneous
+      const partitionedLines = partition(lines, (line: string) => isExtraneousReference(masterDependencies, line))
+      const [extraneousTypings, usedTypings] = partitionedLines
+
+      // Write out references that are being pruned
+      extraneousTypings.forEach((line: string) => emitter.emit('pruneReference', dtsPath, line, getReferenceType(line)))
```

Comment:
Do you want the same event for pruning the d.ts files versus the typings folders?